### PR TITLE
Fix UTF8 string marshalling regression

### LINKED ISSUE #69349: Interop\StringMarshalling\UTF8\UTF8Test failing in outerloop
Pipeline run: https://dev.azure.com/dnceng/public/_build/results?buildId=1771054
Example failure:
```scala
  Starting:    Interop.StringMarshalling.XUnitWrapper (parallel test collections = on, max threads = 2)
    Interop/StringMarshalling/UTF8/UTF8Test/UTF8Test.sh [FAIL]
      free(): invalid pointer
      /datadisks/disk1/work/B90F09CF/w/B12609F6/e/Interop/StringMarshalling/UTF8/UTF8Test/UTF8Test.sh: line 382:  8806 Aborted                 (core dumped) $LAUNCHER $ExePath "${CLRTestExecutionArguments[@]}"
      
      Return code:      1
      Raw output file:      /datadisks/disk1/work/B90F09CF/w/B12609F6/uploads/Reports/Interop.StringMarshalling/UTF8/UTF8Test/UTF8Test.output.txt
      Raw output:
      BEGIN EXECUTION
      /datadisks/disk1/work/B90F09CF/p/corerun -p System.Reflection.Metadata.MetadataUpdater.IsSupported=false UTF8Test.dll ''
      [createdump] Gathering state for process 8806 corerun
      [createdump] Crashing thread 00002266 signal 00000006
      [createdump] Writing minidump with heap to file /home/helixbot/dotnetbuild/dumps/coredump.8806.dmp
      [createdump] Written 129400832 bytes (31592 pages) to core file
      [createdump] Dump successfully written
      Expected: 100
      Actual: 134
      END EXECUTION - FAILED
      Test Harness Exitcode is : 1
      To run the test:
      > set CORE_ROOT=/datadisks/disk1/work/B90F09CF/p
      > /datadisks/disk1/work/B90F09CF/w/B12609F6/e/Interop/StringMarshalling/UTF8/UTF8Test/UTF8Test.sh
      Expected: True
      Actual:   False
      Stack Trace:
           at Interop_StringMarshalling._UTF8_UTF8Test_UTF8Test_._UTF8_UTF8Test_UTF8Test_sh()
```

### PR DESCRIPTION
**Pull request by jkotas**
We need to compensate for the differences in lifetime management patterns used by built-in marshalling system vs. the publicly explosed marshallers.

Fixes #69349

### CODE REVIEW

**Comment by AaronRobinsonMSFT** [comment]
Failure on Linux ARM is

```
Starting:    tracing.eventpipe.XUnitWrapper (parallel test collections = on, max threads = 4)
    tracing/eventpipe/eventsvalidation/ExceptionThrown_V1/ExceptionThrown_V1.sh [FAIL]
      
      Assert failure(PID 5146 [0x0000141a], Thread: 5148 [0x141c]): codeLength > 0
          File: /__w/1/s/src/coreclr/vm/eetwain.cpp Line: 5880
          Image: /root/helix/work/correlation/corerun
      
      /root/helix/work/workitem/e/tracing/eventpipe/eventsvalidation/ExceptionThrown_V1/ExceptionThrown_V1.sh: line 408:  5146 Aborted                 (core dumped) $LAUNCHER $ExePath "${CLRTestExecutionArguments[@]}"
      
      Return code:      1
      Raw output file:      /root/helix/work/workitem/uploads/Reports/tracing.eventpipe/eventsvalidation/ExceptionThrown_V1/ExceptionThrown_V1.output.txt
      Raw output:
      BEGIN EXECUTION
      /root/helix/work/correlation/corerun -p System.Reflection.Metadata.MetadataUpdater.IsSupported=false ExceptionThrown_V1.dll ''
        0.0s: ==TEST STARTING==
        3.5s: Started sending sentinel events...
        3.6s: Connecting to EventPipe...
        5.2s: Creating EventPipeEventSource...
        5.2s: EventPipeEventSource created
        5.3s: Dynamic.All callback registered
        5.3s: Starting stream processing...
        6.1s: Saw new provider 'Microsoft-DotNETCore-SampleProfiler'
        6.2s: Saw sentinel event
        6.3s: Stopped sending sentinel events
        6.3s: Starting event generating action...
        6.3s: Thrown an excpetion 0 times...
        6.3s: Thrown an excpetion 100 times...
        6.4s: Thrown an excpetion 200 times...
        6.4s: Saw new provider 'Microsoft-Windows-DotNETRuntime'
        6.5s: Thrown an excpetion 300 times...
        6.5s: Thrown an excpetion 400 times...
        6.6s: Thrown an excpetion 500 times...
        6.6s: Thrown an excpetion 600 times...
        6.6s: Thrown an excpetion 700 times...
        6.7s: Thrown an excpetion 800 times...
        6.7s: Thrown an excpetion 900 times...
        6.8s: Stopping event generating action
        6.8s: Sending StopTracing command...
      [createdump] Gathering state for process 5146 corerun
      [createdump] Crashing thread 0000141c signal 00000006
      [createdump] Writing minidump with heap to file /home/helixbot/dotnetbuild/dumps/coredump.5146.dmp
      [createdump] Written 146190336 bytes (35691 pages) to core file
      [createdump] Dump successfully written
      Expected: 100
      Actual: 134
      END EXECUTION - FAILED
      Test Harness Exitcode is : 1
      To run the test:
      > set CORE_ROOT=/root/helix/work/correlation
      > /root/helix/work/workitem/e/tracing/eventpipe/eventsvalidation/ExceptionThrown_V1/ExceptionThrown_V1.sh
      Expected: True
      Actual:   False
      Stack Trace:
           at tracing_eventpipe._eventsvalidation_ExceptionThrown_V1_ExceptionThrown_V1_._eventsvalidation_ExceptionThrown_V1_ExceptionThrown_V1_sh()
      Output:
        
        Assert failure(PID 5146 [0x0000141a], Thread: 5148 [0x141c]): codeLength > 0
            File: /__w/1/s/src/coreclr/vm/eetwain.cpp Line: 5880
            Image: /root/helix/work/correlation/corerun
        
        /root/helix/work/workitem/e/tracing/eventpipe/eventsvalidation/ExceptionThrown_V1/ExceptionThrown_V1.sh: line 408:  5146 Aborted                 (core dumped) $LAUNCHER $ExePath "${CLRTestExecutionArguments[@]}"
        
        Return code:      1
        Raw output file:      /root/helix/work/workitem/uploads/Reports/tracing.eventpipe/eventsvalidation/ExceptionThrown_V1/ExceptionThrown_V1.output.txt
        Raw output:
        BEGIN EXECUTION
        /root/helix/work/correlation/corerun -p System.Reflection.Metadata.MetadataUpdater.IsSupported=false ExceptionThrown_V1.dll ''
          0.0s: ==TEST STARTING==
          3.5s: Started sending sentinel events...
          3.6s: Connecting to EventPipe...
          5.2s: Creating EventPipeEventSource...
          5.2s: EventPipeEventSource created
          5.3s: Dynamic.All callback registered
          5.3s: Starting stream processing...
          6.1s: Saw new provider 'Microsoft-DotNETCore-SampleProfiler'
          6.2s: Saw sentinel event
          6.3s: Stopped sending sentinel events
          6.3s: Starting event generating action...
          6.3s: Thrown an excpetion 0 times...
          6.3s: Thrown an excpetion 100 times...
          6.4s: Thrown an excpetion 200 times...
          6.4s: Saw new provider 'Microsoft-Windows-DotNETRuntime'
          6.5s: Thrown an excpetion 300 times...
          6.5s: Thrown an excpetion 400 times...
          6.6s: Thrown an excpetion 500 times...
          6.6s: Thrown an excpetion 600 times...
          6.6s: Thrown an excpetion 700 times...
          6.7s: Thrown an excpetion 800 times...
          6.7s: Thrown an excpetion 900 times...
          6.8s: Stopping event generating action
          6.8s: Sending StopTracing command...
        [createdump] Gathering state for process 5146 corerun
        [createdump] Crashing thread 0000141c signal 00000006
        [createdump] Writing minidump with heap to file /home/helixbot/dotnetbuild/dumps/coredump.5146.dmp
        [createdump] Written 146190336 bytes (35691 pages) to core file
        [createdump] Dump successfully written
        Expected: 100
        Actual: 134
        END EXECUTION - FAILED
        Test Harness Exitcode is : 1
        To run the test:
        > set CORE_ROOT=/root/helix/work/correlation
        > /root/helix/work/workitem/e/tracing/eventpipe/eventsvalidation/ExceptionThrown_V1/ExceptionThrown_V1.sh
```

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/Utf16StringMarshaller.cs [hunk 1]
```diff
@@ -25,17 +25,7 @@ namespace System.Runtime.InteropServices.Marshalling
...
-                _nativeValue = null;
-                return;
-            }
-
-            // + 1 for null terminator
-            _nativeValue =  (ushort*)Marshal.AllocCoTaskMem((str.Length + 1) * sizeof(ushort));
-
-            str.CopyTo(new Span<char>(_nativeValue, str.Length));
-            _nativeValue[str.Length] = '\0'; // null-terminate
+            _nativeValue = (ushort*)Marshal.StringToCoTaskMemUni(str);
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/Utf16StringMarshaller.cs:28]: Note for other reviewers. The `StringToCoTaskMemUni` logic is almost identical to what is being replaced here.

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Comment by jkotas** [comment]
> Failure on Linux ARM

Opened https://github.com/dotnet/runtime/issues/69375

The other outerloop failure is tracked by #61168

**Pull request merged**